File: src/hao_communication/src/string_client.cpp
/**
 * 该例程将请求print_string服务，std_srvs::SetBool
 */
 
#include "ros/ros.h"
#include "std_srvs/SetBool.h"

int main(int argc, char **argv)
{
    // ROS节点初始化
    ros::init(argc, argv, "string_client");

    // 创建节点句柄
    ros::NodeHandle n;

    // 创建一个client，service消息类型是std_srvs::SetBool
    ros::ServiceClient client = n.serviceClient<std_srvs::SetBool>("print_string");

    // 创建std_srvs::SetBool类型的service消息
    std_srvs::SetBool srv;
    srv.request.data = true;

    // 发布service请求，等待应答结果
    if (client.call(srv))
    {
        ROS_INFO("Response : [%s] %s", srv.response.success?"True":"False", 
									   srv.response.message.c_str());
    }
    else
    {
        ROS_ERROR("Failed to call service print_string");
        return 1;
    }

    return 0;
}

File: homework/learning_communication/scripts/string_client.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# 该例程将请求print_string服务，std_srvs::SetBool

import sys
import rospy
from std_srvs.srv import SetBool, SetBoolRequest

def string_client():
	# ROS节点初始化
    rospy.init_node('string_client')

	# 发现print_string服务后，创建一个服务客户端，连接名为print_string的service
    rospy.wait_for_service('print_string')
    try:
        string_client = rospy.ServiceProxy('print_string', SetBool)

		# 请求服务调用，输入请求数据
        response = string_client(True)
        return response.success, response.message
    except rospy.ServiceException, e:
        print "Service call failed: %s"%e

if __name__ == "__main__":
	#服务调用并显示调用结果
    print "Response : [%s] %s" %(string_client())

